proc_macros/src/lib.rs

```diff
@@ -165,3 +203,599 @@ pub fn rule_tests(input: TokenStream) -> TokenStream {
+        struct #rule_instance_struct_name {
+            rule: std::sync::Arc<#rule_struct_name>,
+            listener_queries: Vec<crate::rule::RuleListenerQuery>,
+            #(#state_field_names: #state_field_types),*
+        }
+    }
+}
+
+fn get_rule_instance_rule_instance_impl(
+    rule_instance_struct_name: &Ident,
+    rule_instance_per_file_struct_name: &Ident,
+    rule_instance_per_file_state_fields: &[&RuleStateFieldSpec],
+) -> proc_macro2::TokenStream {
+    let rule_instance_per_file_state_field_names = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.name);
+    let rule_instance_per_file_state_field_initializers = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| match field.initializer.as_ref() {
+            Some(initializer) => quote!(#initializer),
+            None => quote!(Default::default()),
+        });
+    quote! {
+        impl crate::rule::RuleInstance for #rule_instance_struct_name {
+            fn instantiate_per_file(
+                self: std::sync::Arc<Self>,
+                _file_run_info: &crate::rule::FileRunInfo,
+            ) -> Arc<dyn crate::rule::RuleInstancePerFile> {
+                std::sync::Arc::new(#rule_instance_per_file_struct_name {
+                    rule_instance: self,
+                    #(#rule_instance_per_file_state_field_names: #rule_instance_per_file_state_field_initializers),*
+                })
+            }
+
+            fn rule(&self) -> Arc<dyn crate::rule::Rule> {
+                self.rule.clone()
+            }
+
+            fn listener_queries(&self) -> &[crate::rule::RuleListenerQuery] {
+                &self.listener_queries
+            }
+        }
+    }
+}
+
+fn get_rule_instance_per_file_struct_definition(
+    rule_instance_struct_name: &Ident,
+    rule_instance_per_file_struct_name: &Ident,
+    rule_instance_per_file_state_fields: &[&RuleStateFieldSpec],
+) -> proc_macro2::TokenStream {
+    let state_field_names = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.name);
+    let state_field_types = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.type_);
+    quote! {
+        struct #rule_instance_per_file_struct_name {
+            rule_instance: std::sync::Arc<#rule_instance_struct_name>,
+            #(#state_field_names: #state_field_types),*
+        }
+    }
+}
+
+struct SelfAccessRewriter<'a> {
+    rule: &'a Rule,
+}
+
+impl<'a> visit_mut::VisitMut for SelfAccessRewriter<'a> {
+    fn visit_expr_field_mut(&mut self, node: &mut ExprField) {
+        if let Some(self_field_name) = get_self_field_access_name(node) {
+            match self.rule.get_rule_state_scope_for_field(&self_field_name) {
+                Some(RuleStateScope::RuleStatic) => {
+                    let self_field_name = format_ident!("{}", self_field_name);
+                    *node = parse_quote!(self.rule_instance.rule.#self_field_name);
```

Comment:
This isn't totally fleshed out, this seems like it's hard-coding the idea that it's being invoked from the "context" of the `RuleInstancePerFile`-implementing struct, but I think it should be more general than that and also be able to do `self.whatever` re-writings if invoked against the `RuleInstance`-implementing struct (where there the only rewriting that could happen would be if it encounters a "rule-static" field reference then it would rewrite that from `self.whatever` -> `self.rule.whatever`